### PR TITLE
Match page

### PR DESCRIPTION
**Pull request by ofekashery**

### CODE REVIEW

**Comment by fangeugene** [comment]
Dang, the GitHub Actions tests aren't running because it's still in beta and you don't have it :(

**Comment by ofekashery** [comment]
I'm on the waitlist🙃

File: src/components/BreakdownRow/index.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  blue: {
+    backgroundColor: "#0000ff22",
+  },
+  total: {
+    fontWeight: "bold",
+  },
+});
+
+const BreakdownRow = ({
+  data,
```
Comment (fangeugene) [inline src/components/BreakdownRow/index.js]: Is there a reason you're passing `data` in as an array, as opposed to breaking up the label and red/blue values to be more verbose about it?

File: src/components/MatchBreakdownTable/breakdown2019.js [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+          )),
+        ]}
+        vertical={true}
+        subtotal={true}
+      />
+
+      <BreakdownRow
+        data={[
+          "Robot 1 Sandstorm Bonus",
+          getSandstormBonusFor(redBreakdown, 1),
```
Comment (fangeugene) [inline src/components/MatchBreakdownTable/breakdown2019.js]: This `get*For` pattern isn't very "React." I guess `getSandstormBonusFor` is fine since it's just a string, but for things like `getCargoShipDataFor` it's awkward and problematic. See below.

File: src/components/MatchBreakdownTable/breakdown2019.js [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+  const result = breakdown["endgameRobot" + robotNumber];
+  if (result.includes("HabLevel")) {
+    const level = result.substr(-1);
+    const climbPoints = [3, 6, 12];
+    return `Level ${level} (+${climbPoints[level - 1]})`;
+  }
+  return xIcon;
+};
+
+const getCargoShipDataFor = breakdown => {
```
Comment (fangeugene) [inline src/components/MatchBreakdownTable/breakdown2019.js]: `getCargoShipDataFor(breakdown)` really should just be `<CargoShipData breakdown={breakdoown} />`, and instead of returning `[ImageCount(), ImageCount(), ImageCount()]` it should return

```
<>
  <ImageCount />
  <ImageCount />
  <ImageCount />
</>
```

Right now the console has a warning `Warning: Each child in a list should have a unique "key" prop.` because you'r returning an array.

TLDR: While React components are technically functions, composing them as components rather than function calls is preferred.

File: src/components/MatchBreakdownTable/index.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+const MatchBreakdownTable = ({ match }) => {
+  if (match.getYear() === 2019) {
+    return <MatchBreakdown2019 match={match} />;
+  }
+  return <h4>No Breakdown Found</h4>;
+};
+
+MatchBreakdownTable.propTypes = {
+  match: PropTypes.object.isRequired,
```
Comment (fangeugene) [inline src/components/MatchBreakdownTable/index.js]: Can you change this to `PropTypes.instanceOf(Match)` and `import Match from "../../database/Match"

File: src/components/MatchVideos/index.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+              title="Match Video"
+              className={classes.embedIframe}
+              src={`https://www.youtube.com/embed/${video.get("key")}`}
+              frameBorder="0"
+              allowFullScreen
+            />
+          </div>
+        );
+      })}
+      {event != null && (
```
Comment (fangeugene) [inline src/components/MatchVideos/index.js]: Why does `event != null` need to be checked here?

File: src/components/MatchVideos/index.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        <h3>
+          <Button variant="outlined" color="primary" className={classes.button}>
```
Comment (fangeugene) [inline src/components/MatchVideos/index.js]: Let's remove this button for now.

File: src/database/Match.js [hunk 7]
```diff
@@ -47,6 +47,22 @@ export default class Match extends Record({
     }
   }
 
+  getShortName() {
```
Comment (fangeugene) [inline src/database/Match.js]: You can use `getDisplayName(short=true)` instead.

File: src/components/MatchVideos/index.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  },
+  button: {
+    marginRight: theme.spacing(1),
+  },
+}));
+
+const MatchVideos = ({ match, event }) => {
+  const classes = useStyles();
+  const youtubeSearchLink = `https://www.youtube.com/results?search_query=${encodeURIComponent(
+    `${match.getShortName()} ${event.year} ${event.name}`
```
Comment (fangeugene) [inline src/components/MatchVideos/index.js]: Use `getDisplayName(short=true)`

File: src/pages/match.js [hunk 9]
```diff
@@ -1,24 +1,100 @@
...
+      } FIRST Robotics Competition in ${event.getCityStateCountry()}.`}
+      isLoading={
+        matchFetchStatus === "fetching" || eventFetchStatus === "fetching"
+      }
+      refreshFunction={handleRefresh}
+    >
+      <Typography variant="h4" gutterBottom>
+        {match.getDisplayName()} @{" "}
+        <small>
+          {match && <Link href={`/event/${eventKey}`}>{event.name}</Link>}
```
Comment (fangeugene) [inline src/pages/match.js]: Should be ```<Link href={`/event?eventKey=${eventKey}`} as={`/event/${eventKey}`}>```.
Silly, I know. I don't like it either.

File: src/pages/match.js [hunk 10]
```diff
@@ -1,24 +1,100 @@
...
+        </small>
+      </Typography>
+      <Grid container>
+        <Grid item xs={12} md={6} style={{ padding: 8 }}>
```
Comment (fangeugene) [inline src/pages/match.js]: The Grid container has spacing options you can use instead of `padding`

https://material-ui.com/api/grid/

Comment (ofekashery) [reply src/pages/match.js]: A horizontal scroll will appear if we use the ``spacing``.
https://material-ui.com/components/grid/#negative-margin

File: src/components/MatchVideos/index.js (hunk 8 above)
Comment (ofekashery) [reply src/components/MatchVideos/index.js]: It's shorter than that.
For example:
``getDisplayName`` => ``QM 1``
``getShortName`` => ``Q1``
This is what the site uses for the YouTube searching.

File: src/database/Match.js (hunk 7 above)
Comment (ofekashery) [reply src/database/Match.js]: See below.

File: src/components/MatchVideos/index.js (hunk 8 above)
Comment (fangeugene) [reply src/components/MatchVideos/index.js]: Actually `getDisplayName(true)` => `Quals 1`, but yeah I see now that they are different.

That being said, I don't think there is a particular reason why that is why the old site uses this format for YouTube searching. I think it was designed before we had any clear naming scheme in mind. Using `getDisplayName()` (the long version) might actually increase searchability. Thoughts?

File: src/pages/match.js [hunk 11]
```diff
@@ -1,24 +1,104 @@
...
+      isLoading={
+        matchFetchStatus === "fetching" || eventFetchStatus === "fetching"
+      }
+      refreshFunction={handleRefresh}
+    >
+      <Typography variant="h4" gutterBottom>
+        {match.getDisplayName()} @{" "}
+        <small>
+          {match && (
+            <Link href={`/event/${eventKey}`} as={`/event/${eventKey}`}>
```
Comment (fangeugene) [inline src/pages/match.js]: This is still wrong it should be ```<Link href={`/event?eventKey=${eventKey}`} as={`/event/${eventKey}`}>```

File: src/components/MatchVideos/index.js (hunk 8 above)
Comment (ofekashery) [reply src/components/MatchVideos/index.js]: Sounds good

**Comment by fangeugene** [comment]
Can you pull the event page stuff out of this PR? It makes it harder to review. Thanks.

**Comment by ofekashery** [comment]
@fangeugene Done

File: src/components/PageTabs/index.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
+import React, { PureComponent } from "react";
```
Comment (fangeugene) [inline src/components/PageTabs/index.js]: This file doesn't need to be here.

File: src/components/MatchBreakdownTable/breakdown2019.js (hunk 3 above)
Comment (fangeugene) [reply src/components/MatchBreakdownTable/breakdown2019.js]: Can you fix this? The console is throwing warnings still.

File: src/components/MatchBreakdownTable/breakdown2019.js (hunk 2 above)
Comment (fangeugene) [reply src/components/MatchBreakdownTable/breakdown2019.js]: Per the discussion in #dev-ios in Slack, can you change this to not pass the data as an array, but rather as separate props?

File: src/components/BreakdownRow/index.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    </Grid>
+  );
+};
+
+BreakdownRow.propTypes = {
+  label: PropTypes.string.isRequired,
+  red: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.number,
+    PropTypes.object,
```
Comment (fangeugene) [inline src/components/BreakdownRow/index.js]: This should be `PropTypes.node`

**Comment by fangeugene** [comment]
Hurrah! :confetti_ball: Nice work!

**Comment by ofekashery** [comment]
Great!! I'll start working on the event page.
